### PR TITLE
Add brute force protection for login

### PR DESCRIPTION
**Pull request by butlerx**
Account gets locked after number of failed logins thats specified in the config, default 3. The lock is reset after the user changes the password or by calling the set_user_lock and changing the value.

### CODE REVIEW

**Comment by thekemkid** [comment]
Two minor suggestions: It might be nice to add a lock timeout to this (e.g. you can try again after an hour) and give devs the ability to remove the brute force protection (e.g. if locktries is set to null there is no lock protection added)

**Comment by Wardormeur** [comment]
Timeout requires us an extra-work as an integration with node-cron. We considered the retrieval process is sortof related to your own workflow and shouldn't be enforced, likely what has been done with mailing on seneca-user. In our case @Coderdojo, we prefer an email reset, which wouldn't be implemented here ( seneca-user ) but through our own µs, cd-users, since seneca-user doesn't do emailing anymore.
tldr : We're happy to participate to seneca's growth, but we don't have the workforce to develop feature we won't use, alas :)

**Comment by butlerx** [comment]
ive added the null check to disable it

**Comment by thekemkid** [comment]
@Wardormeur Fair enough, that's a valid reason 👍

I'm not the maintainer of this, so Its up to @mihaidma or @mirceaalexandru to merge, but this lgtm :)

**Comment by mirceaalexandru** [comment]
Hey @Wardormeur, 
first great job, thank you for this PR.

Now I do have two comments:
- I would like this to be  disabled by default as we do want to preserve the current functionality in case that somebody use the new version, don't see this feature and did not implement the reset/notify using email. I do not want them to have locked accounts. For this reason I think is best to have this feature turn off by default.
- This will require, I think, a major version change. This is because you added a new property to user and on some DB engines this means to change the sys_user table structure. Because of this I think we should consider this as a breaking change.

@mihaidma any thoughts?

File: user.js [hunk 1]
```diff
@@ -661,13 +675,49 @@ module.exports = function user (options) {
...
+  // - failure: {ok:false, why:}
+  function cmd_set_user_lock (args, done) {
+    var seneca = this
+    var why
+    var userent = seneca.make(user_canon)
+    var ok
+
+    userent.load$({id: args.id}, function (err, user) {
+      if (err) return done(err)
+      console.log(user.lockTry, args);
```
Comment (mirceaalexandru) [inline user.js]: Please do not use console.log, use seneca.log

File: user.js [hunk 2]
```diff
@@ -661,13 +675,49 @@ module.exports = function user (options) {
...
+  // - success: {ok:true, why:}
+  // - failure: {ok:false, why:}
+  function cmd_set_user_lock (args, done) {
+    var seneca = this
+    var why
+    var userent = seneca.make(user_canon)
+    var ok
+
+    userent.load$({id: args.id}, function (err, user) {
+      if (err) return done(err)
```
Comment (mirceaalexandru) [inline user.js]: this done(err) do not match the description of the function that say returns `{ok: false, why: err}`

File: user.js [hunk 3]
```diff
@@ -661,13 +675,49 @@ module.exports = function user (options) {
...
+        ok = false
+        why = args.why
+        console.log(user.lockTry);
+      } else {
+        user.lockTry = 0
+        ok = true
+        why = 'account-unlocked'
+      }
+      user.save$(function (err, user) {
+        if (err) return done(err)
```
Comment (mirceaalexandru) [inline user.js]: same as above (`done(err)`)

**Comment by mirceaalexandru** [comment]
One more note: can you please add to this PR some tests? We need tests for all new features.

File: user.js (hunk 3 above)
Comment (Wardormeur) [inline user.js]: and console log ;)
I'd even say you could += instead of x = x +1, but that's just me :)

File: user.js (hunk 2 above)
Comment (Wardormeur) [inline user.js]: not sure about seneca coding style, but i'd prefer ```var x, y ,z ``` instead of  ``` var x vary var z```

**Comment by mirceaalexandru** [comment]
Hey @butlerx really sorry, I just realize is your PR actually. :+1:

File: user.js (hunk 2 above)
Comment (mirceaalexandru) [inline user.js]: @Wardormeur well here the coding standard requires one var/line so its OK like it is now.

**Comment by mirceaalexandru** [comment]
It seems to me that after a successful login the counter should be reset to 0. 

This seems to be the desired feature. If you try x times to login without success the account is locked. But if I failed 1 time and then login with success it should reset the counter back to 0.

Only x consecutive logins should lock the account. 

@butlerx @Wardormeur @mihaidma @thekemkid Thoughts?

**Comment by butlerx** [comment]
yep thats what i was goign for and a reset to the user password would unlock the account or by calling the unlock. ill get on to the other parts now aswell

File: user.js (hunk 1 above)
Comment (butlerx) [inline user.js]: i actually meant to remove those sorry

**Comment by Wardormeur** [comment]
Very neat to see that going, keep up @butlerx  :)

**Comment by butlerx** [comment]
tests added and disabled by default
it does require a modification to user so would require a migration for dbs so the version number is up too you

**Comment by mcdonnelldean** [comment]
LGTM @mirceaalexandru / @mihaidma I'd say major version to be sure?

**Comment by mcdonnelldean** [comment]
@mihaidma Can we review this and get it in?

File: README.md [hunk 4]
```diff
@@ -430,6 +431,19 @@ Object with properties:
 
    * `ok`: true if operation is OK
 
+### role:user, cmd:set_user_lock
+
+Sets the lock on users account
+
+#### Argumants:
```
Comment (mihaidma) [inline README.md]: typo

File: user.js [hunk 5]
```diff
@@ -526,6 +533,7 @@ module.exports = function user (options) {
     user.name = args.name || ''
     user.active = void 0 === args.active ? true : args.active
     user.when = new Date().toISOString()
+    user.lockTry = 0;
```
Comment (mihaidma) [inline user.js]: linting will fail with this

File: README.md [hunk 6]
```diff
@@ -430,6 +431,19 @@ Object with properties:
...
+
+   * `id`: the id of the user
+   * `failTry`: boolean to say if the login failed or not
+   * `why`: why the accouunts lock status is being changed
+
+#### Provides:
+
+   * `ok`: true if the account was unloocked or false if the lock was incremented
```
Comment (mihaidma) [inline README.md]: typo unlooked

File: test/lock-unlock.test.js [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+    si.act({role: 'user', cmd: 'login', nick: user2Data.nick, password: user2Data.password}, function (err, data) {
+      expect(err).to.not.exist()
+      expect(data.ok).to.be.true()
+      user2Id = data.user.id
+      done(err)
+    })
+  })
+
+  for(var i=0; i<3; i++) {
+    test('user/login user test', function (done) {
```
Comment (mihaidma) [inline test/lock-unlock.test.js]: this loop makes and attempt to login with wrong password for user1 then checks user lock with user2ID? Seems incorrect.
Also these tests should be grouped under the same test.

Comment (mihaidma) [inline test/lock-unlock.test.js]: Test is ok-ish, I got tricked by the test arrangement.

**Comment by mihaidma** [comment]
I'll merge, the implemented functionality is ok. After merge will do the fixes asked in comments.
i'll not publish yet the plugin, planning to review the set_user_lock 'interface', it might change

**Comment by mihaidma** [comment]
I made the specs for brute force protection: https://github.com/senecajs/seneca-user/issues/62
Your implementation remains more or less the same, the changes are:
- renamed the value stored in db,
- the set_usor_lock command is not exposed,
- only the unlock functionality is exposed

I'll work at modifying this implementation.

**Comment by Wardormeur** [comment]
Very nice, thanks both of you :)

**Comment by mcdonnelldean** [comment]
Good work @butlerx and @mihaidma ❤️
